packages/firestore/src/local/indexeddb_query_cache.ts

```diff
@@ -97,36 +97,75 @@ export class IndexedDbQueryCache implements QueryCache {
     );
   }
 
-  addQueryData(
+  private saveQueryData(
     transaction: PersistenceTransaction,
     queryData: QueryData
   ): PersistencePromise<void> {
-    const targetId = queryData.targetId;
-    const addedQueryPromise = targetsStore(transaction).put(
-      this.serializer.toDbTarget(queryData)
-    );
-    if (targetId > this.metadata.highestTargetId) {
-      this.metadata.highestTargetId = targetId;
-      return addedQueryPromise.next(() =>
-        globalTargetStore(transaction).put(DbTargetGlobal.key, this.metadata)
-      );
-    } else {
-      return addedQueryPromise;
+    return targetsStore(transaction).put(this.serializer.toDbTarget(queryData));
+  }
+
+  // Updates the in-memory version of the metadata. Saving is done separately.
+  // Returns true if there were any changes to the metadata.
+  private updateMetadata(queryData: QueryData): boolean {
+    let needsUpdate = false;
+    if (queryData.targetId > this.metadata.highestTargetId) {
+      this.metadata.highestTargetId = queryData.targetId;
+      needsUpdate = true;
     }
+
+    // TODO(gsoltis): add sequence number check
+    return needsUpdate;
   }
 
-  removeQueryData(
+  private saveMetadata(
+    transaction: PersistenceTransaction
+  ): PersistencePromise<void> {
+    return globalTargetStore(transaction).put(
+      DbTargetGlobal.key,
+      this.metadata
+    );
+  }
+
+  addQueryData(
+    transaction: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    return this.saveQueryData(transaction, queryData).next(() => {
+      this.metadata.targetCount += 1;
+      this.updateMetadata(queryData);
+      return this.saveMetadata(transaction);
+    });
+  }
+
+  updateQueryData(
     transaction: PersistenceTransaction,
     queryData: QueryData
   ): PersistencePromise<void> {
-    return this.removeMatchingKeysForTargetId(
-      transaction,
-      queryData.targetId
-    ).next(() => {
-      targetsStore(transaction).delete(queryData.targetId);
+    return this.saveQueryData(transaction, queryData).next(() => {
+      if (this.updateMetadata(queryData)) {
+        return this.saveMetadata(transaction);
+      } else {
+        return PersistencePromise.resolve();
+      }
     });
   }
 
+  removeQueryData(
+    transaction: PersistenceTransaction,
+    queryData: QueryData
+  ): PersistencePromise<void> {
+    return this.removeMatchingKeysForTargetId(transaction, queryData.targetId)
+      .next(() => targetsStore(transaction).delete(queryData.targetId))
+      .next(() => {
+        this.metadata.targetCount -= 1;
```

Comment:
The comment for removeQueryData() says "no-op if no entry exists".  We need to either change the comment (and add appropriate assertions / comments about existence requirement), or change the implementation to check for existence first...

As a sanity check, we may want to add a targetCount >= 0 assertion here.

Reply:
Comment changed.